lib/types.nix

```diff
@@ -367,22 +367,47 @@ rec {
         name = "coercedTo";
         description = "${finalType.description} or ${coercedType.description}";
         check = x: finalType.check x || coercedType.check x;
-        merge = loc: defs:
+        merge =
           let
             coerceVal = val:
               if finalType.check val then val
               else let
                 coerced = coerceFunc val;
               in assert finalType.check coerced; coerced;
 
-          in finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
+          in loc: defs: finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
         getSubOptions = finalType.getSubOptions;
         getSubModules = finalType.getSubModules;
         substSubModules = m: coercedTo coercedType coerceFunc (finalType.substSubModules m);
         typeMerge = t1: t2: null;
         functor = (defaultFunctor name) // { wrapped = finalType; };
       };
 
+    # Function from `from` to `to`.
+    function = from: to:
+      assert from.getSubModules == null;
+      mkOptionType rec {
+        name = "function";
```

Comment:
Maybe we could have a `pkgs`-like module argument which corresponds to the `recursiveUpdate` of both `pkgs` and `pkgsi686Linux` packages, with a predicate to stop at derivations.  Thus selecting a attribute name in the zipped set of packages would select the same attribute name in both sets.